Expose all permits acquisition in IndexShard and TransportReplicationAction

tlrx opened the pull request:
This pull request exposes two new methods in the `IndexShard` and `TransportReplicationAction` classes in order to allow transport replication actions to acquire all index shard operation permits for their execution.

It first adds the `acquirePrimaryAllOperationsPermits()` and the `acquireReplicaAllOperationsPermits()` methods to the `IndexShard` class which allow to acquire all operations permits on a shard  while exposing a `Releasable`. It also refactors the `TransportReplicationAction` class to expose two protected methods (`acquirePrimaryOperationPermit()` and `acquireReplicaOperationPermit()`) that can be overridden when a transport replication action requires the acquisition of all permits on primary and/or replica shard during execution.

Finally, it adds a `TransportReplicationAllPermitsAcquisitionTests` which illustrates how a transport replication action can grab all permits before adding a cluster block in the cluster state, making subsequent operations that requires a single permit to fail (such test has been discussed in https://github.com/elastic/elasticsearch/pull/35332#discussion_r231891192).

Related to elastic #33888

elasticmachine commented:
Pinging @elastic/es-distributed

bleskes commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2352,58 +2343,121 @@ public void acquireReplicaOperationPermit(final long opPrimaryTerm, final long g
+     * an {@link IllegalStateException}. Otherwise the global checkpoint and the max_seq_no_of_updates marker of the replica are updated
+     * before the invocation of the {@link ActionListener#onResponse(Object)}} method of the provided listener.
+     */
+    private ActionListener<Releasable> createListener(final ActionListener<Releasable> listener,
```
I'm not so comfortable with separating this code from the one in `updatePrimaryTermIfNeeded` - they are tightly connected. Instead of sharing code this way, how about creating a callback that will run:
```
indexShardOperationPermits.acquire(listener, executorOnDelay, true, debugInfo);
```
or 
```
indexShardOperationPermits.asyncBlockOperations(listener, timeout.duration(), timeout.timeUnit());
```

tlrx replied:
I share your concerns and I updated the code, let me know what you think.

tlrx commented on server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -334,7 +334,38 @@ public void messageReceived(ConcreteShardRequest<Request> request, TransportChan
+                    primaryTerm, actualTerm);
+            }
+
+            final ActionListener<PrimaryShardReference> onReferenceAcquired = this;
```
Sure, I pushed 21dac9e4d8e41fa85f5b21159277c23430d2a98a

tlrx commented on server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -697,7 +728,7 @@ public void onFailure(Exception e) {
     }
 
-    protected IndexShard getIndexShard(ShardId shardId) {
+    protected IndexShard getIndexShard(final ShardId shardId, final String targetAllocationID) {
```
This is a leftover that was helpful in `TransportReplicationAllPermitsAcquisitionTests` to dinstinguish between the primary and replica shard when executing a transport replication test action.

I reverted this and changed how the action in test retrieves the index shard.

tlrx commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2281,7 +2281,18 @@ public void acquirePrimaryOperationPermit(ActionListener<Releasable> onPermitAcq
+     * Acquire all primary operation permits. Once all permits are acquired, the provided ActionListener is called.
+     * It is the responsibility of the caller to close the {@link Releasable}.
+     */
+    public void acquirePrimaryAllOperationsPermits(final ActionListener<Releasable> onPermitAcquired, final TimeValue timeout) {
```
Done

tlrx commented:
Thanks @bleskes, I addressed your first bunch of comments. I also added comments to the `TransportReplicationAllPermitsAcquisitionTests` and reworked it a bit so that it is easier to understand.

ywelsch commented on server/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationAllPermitsAcquisitionTests.java:
```diff
@@ -0,0 +1,561 @@
+                allPermitsAction.new AsyncPrimaryAction(request(), allocationId(), primaryTerm(), transportChannel(allPermitFuture), null) {
+                    @Override
+                    void runWithPrimaryShardReference(final TransportReplicationAction.PrimaryShardReference reference) {
+                        assertEquals("All permits must be acquired", 0, reference.indexShard.getActiveOperationsCount());
```
sorry, I missed the ping here. I think that 1 is ok too, but then we should in general change the terminology in `IndexShardOperationPermits` to move from the notion of blocking operations to running exclusive operations (similar to a read/write lock).

bleskes replied:
agreed.
